packages/core/src/routes/admin-user-role.ts

```diff
@@ -114,12 +115,15 @@ export default function adminUserRoleRoutes<T extends AuthedRouter>(
     async (ctx, next) => {
       const {
         params: { userId },
-        body: { roleIds },
+        body: { roleIds: originalRoleIds },
```

Comment:
I am thinking maybe you could write a middleware for these apis, this middleware do the following things:

1. query roles by id
2. assert all roles type is RoleType.User
3. append the roles to a context, so it can be directly retrieved by the route body.

Reply:
Good suggestion, add an issue (LOG-7006) to track the refactor. Let's implement the logic first.